### PR TITLE
Add evaluate logic for text scanner and fix partition column not return error

### LINKED ISSUE #2783: Query from hive external table with CSV format return error
### Steps to reproduce the behavior (Required)
1. create hive partition table, and insert some data:
```
create table hive_hdfs_par_csv(
 col_tinyint tinyint comment "column tinyint"
,col_int int comment "column int"
,col_integer integer comment "column integer"
,col_bigint bigint comment "column bigint"
,col_float float comment "column float"
,col_double double comment "column double"
,col_double_precision double precision comment "column double precision"
,col_decimal decimal comment "column decimal"
,col_timestamp timestamp comment "column timestamp"
,col_string string comment "column string"
,col_varchar varchar(100) comment "column varchar"
,col_binary binary comment "column binary"
,col_char char(100) comment "column char"
,col_boolean boolean comment "column boolean"
)
partitioned by(col_date date,col_smallint smallint)
clustered by(col_int) sorted by(col_int) into 32 buckets
row format delimited fields terminated by ','
stored as textfile;
insert into hive_hdfs_par_csv partition(col_date = '2021-01-01',col_smallint=1) select 1,1,1,1,1.001,10.1,110.1,1110.1,'2021-10-30 12:10:23','hello world','hi hello','1110001010101011001001','nihao',true;
insert into hive_hdfs_par_csv partition(col_date = '2021-02-01',col_smallint=2) select 2,2,2,2,2.001,20.1,210.1,2110.2,'2021-02-01 12:10:23','hello world','hi hello','1110001010101011001001','nihao',true;
 insert into hive_hdfs_par_csv partition(col_date = '2021-03-01',col_smallint=3) select 3,3,3,3,3.001,30.1,310.1,3110.1,'2021-03-01 12:10:23','hello world','hi abc','1110001010101011001001','nihao',true;
 insert into hive_hdfs_par_csv partition(col_date = '2021-04-01',col_smallint=4) select 4,4,4,4,4.001,40.1,410.1,4110.2,'2021-04-01 12:10:23','hello world','hi','1110001010101011001001','nihao',true;
 insert into hive_hdfs_par_csv partition(col_date = '2021-05-01',col_smallint=5) select 5,5,5,5,5.001,50.1,510.1,5110.1,'2021-05-01 12:10:23','hello world','hi','1110001010101011001001','nihao',true;
```

2. create hive external table:
```
create external table ex_hive_par_csv (              
 col_tinyint tinyint null comment "column tinyint"             
 ,col_int int comment "column int"             
 ,col_integer integer comment "column integer"             
 ,col_bigint bigint comment "column bigint"             
 ,col_float float comment "column float"             
 ,col_double double comment "column double"             
 ,col_double_precision double comment "column double precision"             
 ,col_decimal decimal comment "column decimal"             
 ,col_timestamp datetime comment "column timestamp"             
 ,col_string varchar(1000) comment "column string"             
 ,col_varchar varchar(5) comment "column varchar"             
 ,col_binary varchar(1000) comment "column binary"             
 ,col_char char(100) comment "column char"             
 ,col_boolean boolean null comment "column boolean"             
 ,col_date date comment "column date"             
 ,col_smallint smallint comment "column smallint")         
 ENGINE=hive         
 properties (             
  "resource" = "hive2",             
  "table" = "hive_hdfs_par_csv",             
  "database" = "hive_table_test");
```

3. select from hive external table:
```
mysql> select * from ex_hive_par_csv;
ERROR 1064 (HY000): mismatched row count: 0 vs 1
mysql> select col_tinyint,col_smallint,col_date,col_boolean from ex_hive_par_csv order by col_tinyint
desc;
+-------------+--------------+------------+-------------+
| col_tinyint | col_smallint | col_date   | col_boolean |
+-------------+--------------+------------+-------------+
|           5 |            0 | �193-02-17  |           1 |
|           4 |         8192 | }901-05-17 |           1 |
|           3 |       -32768 | �929-06-18  |           1 |
|           2 |       -32768 | �268-03-10  |           1 |
|           1 |            0 | �631-05-26  |           1 |
+-------------+--------------+------------+-------------+
5 rows in set (0.01 sec)

mysql> select col_date from ex_hive_par_csv;
Empty set (0.02 sec)

mysql> select col_smallint from ex_hive_par_csv;
Empty set (0.03 sec)

mysql> select col_varchar from ex_hive_par_csv;
+-------------+
| col_varchar |
+-------------+
| hi          |
| hi          |
| NULL        |
| NULL        |
| NULL        |
+-------------+
5 rows in set (0.01 sec)
```

### Expected behavior (Required)
query from hive external table return right result

### Real behavior (Required)
error or wrong result

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
```
mysql> select current_version();
+------------------------+
| current_version()      |
+------------------------+
| QA_TEST_MASTER f163676 |
+------------------------+
1 row in set (0.00 sec)
```

### LINKED ISSUE #2791: Wrong result returned from query of hive external table with csv format
### Steps to reproduce the behavior (Required)
1. create hive table and insert data:
```
 create table hive_hdfs_csv(
  col_tinyint tinyint comment "column tinyint"
 ,col_smallint smallint comment "column smallint"
 ,col_int int comment "column int"
 ,col_integer integer comment "column integer"
 ,col_bigint bigint comment "column bigint"
 ,col_float float comment "column float"
 ,col_double double comment "column double"
 ,col_double_precision double precision comment "column double precision"
 ,col_decimal decimal comment "column decimal"
 ,col_timestamp timestamp comment "column timestamp"
 ,col_date date comment "column date"
 ,col_string string comment "column string"
 ,col_varchar varchar(100) comment "column varchar"
 ,col_binary binary comment "column binary"
 ,col_char char(100) comment "column char"
 ,col_boolean boolean comment "column boolean"
 ,col_array array<int> comment "column array"
 )
 clustered by(col_int) sorted by(col_int) into 32 buckets
 row format delimited fields terminated by ','
 stored as textfile;
 insert into hive_hdfs_csv select 1,1,1,1,1,1.001,10.1,110.1,1110.1,'2021-10-30 12:10:23','2021-10-30','hello world','hi hello','1110001010101011001001','nihao',true,array(1,10,100);
 insert into hive_hdfs_csv select 2,2,2,2,2,2.001,20.1,210.1,2110.2,'2021-02-01 12:10:23','2021-02-01','hello world','hi hello','1110001010101011001001','nihao',true,array(2,20,200);
 insert into hive_hdfs_csv select 3,3,3,3,3,3.001,30.1,310.1,3110.1,'2021-03-01 12:10:23','2021-03-01','hello world','hi abc','1110001010101011001001','nihao',true,array(3,30,300);
 insert into hive_hdfs_csv select 4,4,4,4,4,4.001,40.1,410.1,4110.2,'2021-04-01 12:10:23','2021-04-01','hello world','hi','1110001010101011001001','nihao',true,array(4,40,400);
 insert into hive_hdfs_csv select 5,5,5,5,5,5.001,50.1,510.1,5110.1,'2021-05-01 12:10:23','2021-05-01','hello world','hi','1110001010101011001001','nihao',true,array(5,50,500);
```
2. create hive external table:
```
create external table ex_hive_csv (              
 col_tinyint tinyint comment "column tinyint"
,col_smallint smallint comment "column smallint"
,col_int int comment "column int"
,col_integer integer comment "column integer"
,col_bigint bigint comment "column bigint"
,col_float float comment "column float"
,col_double double comment "column double"
,col_double_precision double comment "column double precision"
,col_decimal decimal comment "column decimal"
,col_timestamp datetime comment "column timestamp"
,col_date date comment "column date"
,col_string varchar(100) comment "column string"
,col_varchar varchar(100) comment "column varchar"
,col_binary varchar(100) comment "column binary"
,col_char char(100) comment "column char"
,col_boolean boolean comment "column boolean")         
ENGINE=hive         
properties (             
"resource" = "hive2",             
"table" = "hive_hdfs_csv",             
"database" = "hive_table_test");
```
3. select from hive external table:
```
mysql> select * from ex_hive_csv order by 1;
+-------------+--------------+---------+-------------+------------+-----------+------------+----------------------+-------------+---------------------+------------+-------------+-------------+----------------------------------+------------------------------------------------------------------------------------------------------+-------------+
| col_tinyint | col_smallint | col_int | col_integer | col_bigint | col_float | col_double | col_double_precision | col_decimal | col_timestamp       | col_date   | col_string  | col_varchar | col_binary                       | col_char                                                                                             | col_boolean |
+-------------+--------------+---------+-------------+------------+-----------+------------+----------------------+-------------+---------------------+------------+-------------+-------------+----------------------------------+------------------------------------------------------------------------------------------------------+-------------+
|           1 |            1 |       1 |           1 |          1 |     1.001 |       10.1 |                110.1 |        1110 | 2021-10-30 12:10:23 | 2021-10-30 | hello world | hi hello    | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== | nihao                                                                                                |           1 |
|           2 |            2 |       2 |           2 |          2 |     2.001 |       20.1 |                210.1 |        2110 | 2021-02-01 12:10:23 | 2021-02-01 | hello world | hi hello    | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== | nihao                                                                                                |           1 |
|           3 |            3 |       3 |           3 |          3 |     3.001 |       30.1 |                310.1 |        3110 | 2021-03-01 12:10:23 | 2021-03-01 | hello world | hi abc      | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== | nihao                                                                                                |           1 |
|           4 |            4 |       4 |           4 |          4 |     4.001 |       40.1 |                410.1 |        4110 | 2021-04-01 12:10:23 | 2021-04-01 | hello world | hi          | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== | nihao                                                                                                |           1 |
|           5 |            5 |       5 |           5 |          5 |     5.001 |       50.1 |                510.1 |        5110 | 2021-05-01 12:10:23 | 2021-05-01 | hello world | hi          | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== | nihao                                                                                                |           1 |
+-------------+--------------+---------+-------------+------------+-----------+------------+----------------------+-------------+---------------------+------------+-------------+-------------+----------------------------------+------------------------------------------------------------------------------------------------------+-------------+
5 rows in set (0.02 sec)

mysql> select * from ex_hive_csv where col_tinyint = 1;
+-------------+--------------+---------+-------------+------------+-----------+------------+----------------------+-------------+---------------------+------------+-------------+-------------+----------------------------------+------------------------------------------------------------------------------------------------------+-------------+
| col_tinyint | col_smallint | col_int | col_integer | col_bigint | col_float | col_double | col_double_precision | col_decimal | col_timestamp       | col_date   | col_string  | col_varchar | col_binary                       | col_char                                                                                             | col_boolean |
+-------------+--------------+---------+-------------+------------+-----------+------------+----------------------+-------------+---------------------+------------+-------------+-------------+----------------------------------+------------------------------------------------------------------------------------------------------+-------------+
|           1 |            1 |       1 |           1 |          1 |     1.001 |       10.1 |                110.1 |        1110 | 2021-10-30 12:10:23 | 2021-10-30 | hello world | hi hello    | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== | nihao                                                                                                |           1 |
|           2 |            2 |       2 |           2 |          2 |     2.001 |       20.1 |                210.1 |        2110 | 2021-02-01 12:10:23 | 2021-02-01 | hello world | hi hello    | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== | nihao                                                                                                |           1 |
|           3 |            3 |       3 |           3 |          3 |     3.001 |       30.1 |                310.1 |        3110 | 2021-03-01 12:10:23 | 2021-03-01 | hello world | hi abc      | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== | nihao                                                                                                |           1 |
|           5 |            5 |       5 |           5 |          5 |     5.001 |       50.1 |                510.1 |        5110 | 2021-05-01 12:10:23 | 2021-05-01 | hello world | hi          | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== | nihao                                                                                                |           1 |
|           4 |            4 |       4 |           4 |          4 |     4.001 |       40.1 |                410.1 |        4110 | 2021-04-01 12:10:23 | 2021-04-01 | hello world | hi          | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== | nihao                                                                                                |           1 |
+-------------+--------------+---------+-------------+------------+-----------+------------+----------------------+-------------+---------------------+------------+-------------+-------------+----------------------------------+------------------------------------------------------------------------------------------------------+-------------+
5 rows in set (0.03 sec)
```

### Expected behavior (Required)
1 row return with filter: col_tinyint = 1

### Real behavior (Required)
all rows returned

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
```
mysql> select current_version();
+------------------------+
| current_version()      |
+------------------------+
| QA_TEST_MASTER f163676 |
+------------------------+
1 row in set (0.01 sec)
```

### LINKED ISSUE #2839: Query return error when select count(*) from hive external table with csv format
### Steps to reproduce the behavior (Required)
1. create hive table:
```
CREATE EXTERNAL TABLE partsupp (
  ps_partkey int,
  ps_suppkey int,
  ps_availqty int,
  ps_supplycost decimal(15, 2),
  ps_comment varchar(199)
) row format delimited fields terminated by '|'
 stored as textfile
location '/user/hive/warehouse/tpch_csv_1g/partsupp';
```
 
select from hive table OK:
```
hive> select count(*) from partsupp;
WARNING: Hive-on-MR is deprecated in Hive 2 and may not be available in the future versions. Consider using a different execution engine (i.e. spark, tez) or using Hive 1.X releases.
Query ID = dorisdb_20220113162231_66cef907-7157-414e-8a4d-8d12a9fbdbcd
Total jobs = 1
Launching Job 1 out of 1
Number of reduce tasks determined at compile time: 1
In order to change the average load for a reducer (in bytes):
  set hive.exec.reducers.bytes.per.reducer=<number>
In order to limit the maximum number of reducers:
  set hive.exec.reducers.max=<number>
In order to set a constant number of reducers:
  set mapreduce.job.reduces=<number>
Job running in-process (local Hadoop)
2022-01-13 16:22:32,410 Stage-1 map = 100%,  reduce = 100%
Ended Job = job_local1786358483_0009
MapReduce Jobs Launched:
Stage-Stage-1:  HDFS Read: 2423279754 HDFS Write: 0 SUCCESS
Total MapReduce CPU Time Spent: 0 msec
OK
800000
Time taken: 1.27 seconds, Fetched: 1 row(s)
```

2. create hive external table in starrocks cluster:
```
 CREATE external TABLE partsupp (
  ps_partkey int,
  ps_suppkey int,
  ps_availqty int,
  ps_supplycost decimal(15, 2),
  ps_comment varchar(199)
 ) ENGINE=hive         
properties (             
 "resource" = "hive2",             
 "table" = "partsupp",             
 "database" = "tpch_1g_csv");
```
select limit 1 OK:
```
mysql> select * from partsupp limit 1;
+------------+------------+-------------+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| ps_partkey | ps_suppkey | ps_availqty | ps_supplycost | ps_comment                                                                                                                                                                                       |
+------------+------------+-------------+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|          1 |          2 |        3325 |        771.64 | , even theodolites. regular, final theodolites eat after the carefully pending foxes. furiously regular deposits sleep slyly. carefully bold realms above the ironic dependencies haggle careful |
+------------+------------+-------------+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.02 sec)
```

select count(*) return ERROR:
```
mysql> select count(*) from partsupp;
ERROR 1064 (HY000): fail to hdfsPreadFully file, file=hdfs://172.26.92.141:9002/user/hive/warehouse/tpch_csv_1g/partsupp/partsupp.tbl, error=error=Error(255): Unknown error 255, root_cause=EOFException: End of file reached before reading fully.
```

some table with less data select query OK:
```
mysql> select count(*) from part;
+----------+
| count(*) |
+----------+
|   200000 |
+----------+
1 row in set (0.08 sec)

mysql> select count(*) from nation;
+----------+
| count(*) |
+----------+
|       25 |
+----------+
1 row in set (0.01 sec)

mysql> select * from region;
+-------------+-------------+---------------------------------------------------------------------------------------------------------------------+
| r_regionkey | r_name      | r_comment                                                                                                           |
+-------------+-------------+---------------------------------------------------------------------------------------------------------------------+
|           0 | AFRICA      | lar deposits. blithely final packages cajole. regular waters are final requests. regular accounts are according to  |
|           1 | AMERICA     | hs use ironic, even requests. s                                                                                     |
|           2 | ASIA        | ges. thinly even pinto beans ca                                                                                     |
|           3 | EUROPE      | ly final courts cajole furiously final excuse                                                                       |
|           4 | MIDDLE EAST | uickly special accounts cajole carefully blithely close requests. carefully final asymptotes haggle furiousl        |
+-------------+-------------+---------------------------------------------------------------------------------------------------------------------+
5 rows in set (0.01 sec)
```

### Expected behavior (Required)
query return right result

### Real behavior (Required)
query return error

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
```
mysql> select current_version();
+------------------------+
| current_version()      |
+------------------------+
| QA_TEST_MASTER 6bdfbfd |
+------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by caneGuy**
This pr will close #2783 close #2791 close #2839.
Why we resolve these issues in this single pr?
Because we think the three bugs is related with each other.

More details about this change as below:
- Because we do not add project operator follow hdfs scan node, so we need add filter logic into text scanner.This will close #2791 
- For csv or text file we can not select partition column from original file. So we append partition columns manully to fix partition column can not be selected error. This will close #2783 
- Traverse all the scan ranges assigned to the scanner, and check end of file use offset to fix hdfs readfully error of #2839

### CODE REVIEW

File: be/src/exec/vectorized/hdfs_scanner_text.cpp [hunk 1]
```diff
@@ -113,10 +113,25 @@ void HdfsTextScanner::do_close(RuntimeState* runtime_state) noexcept {
 }
 
 Status HdfsTextScanner::do_get_next(RuntimeState* runtime_state, ChunkPtr* chunk) {
-    return _parse_csv(runtime_state->chunk_size(), chunk);
+    CHECK(chunk != nullptr);
+    Status status = parse_csv(runtime_state->chunk_size(), chunk);
+    RETURN_IF_ERROR(status);
```
Comment (imay) [inline be/src/exec/vectorized/hdfs_scanner_text.cpp:118]: how about `RETURN_IF_ERROR(parse_csv(runtime_state->chunk_size(), chunk);`

File: be/src/exec/vectorized/hdfs_scanner_text.cpp [hunk 2]
```diff
@@ -149,17 +164,32 @@ Status HdfsTextScanner::_parse_csv(int chunk_size, ChunkPtr* chunk) {
...
-            if (!_converters[k]->read_string(_column_raw_ptrs[k], field, options)) {
+            if (!_converters[j]->read_string(_column_raw_ptrs[index], field, options)) {
                 chunk->get()->set_num_rows(num_rows);
                 has_error = true;
                 break;
             }
-            k++;
         }
         num_rows += !has_error;
+        if (!has_error) {
```
Comment (imay) [inline be/src/exec/vectorized/hdfs_scanner_text.cpp:178]: Provide some comment to explain this.

File: be/src/exec/vectorized/hdfs_scanner_text.cpp [hunk 3]
```diff
@@ -149,17 +164,32 @@ Status HdfsTextScanner::_parse_csv(int chunk_size, ChunkPtr* chunk) {
...
+            for (int p = 0; p< _file_read_param.partition_columns.size(); ++p) {
```
Comment (imay) [inline be/src/exec/vectorized/hdfs_scanner_text.cpp:179]: better to define a local variable `num_part_columns = _file_read_param.partition_columns.size()`

**Comment by caneGuy** [comment]
cc @dirtysalt @imay PTAL thanks!

File: be/src/exec/vectorized/hdfs_scanner_text.cpp [hunk 4]
```diff
@@ -113,10 +113,24 @@ void HdfsTextScanner::do_close(RuntimeState* runtime_state) noexcept {
...
+
+    ChunkPtr ck = *chunk;
+    // do stats before we filter rows which does not match.
+    _stats.raw_rows_read += ck->num_rows();
+    for (auto& it : _file_read_param.conjunct_ctxs_by_slot) {
+        // do evaluation.
+        SCOPED_RAW_TIMER(&_stats.expr_filter_ns);
+        ExecNode::eval_conjuncts(it.second, ck.get());
+        if (ck->num_rows() == 0) {
+            return Status::OK();
```
Comment (dirtysalt) [inline be/src/exec/vectorized/hdfs_scanner_text.cpp:127]: how about `break`?  so there is only one `return`.

File: be/src/exec/vectorized/hdfs_scanner_text.cpp [hunk 5]
```diff
@@ -149,17 +163,36 @@ Status HdfsTextScanner::_parse_csv(int chunk_size, ChunkPtr* chunk) {
         }
 
         bool has_error = false;
-        for (int j = 0, k = 0; j < _scanner_params.materialize_slots.size(); j++) {
+        int num_materialize_columns = _scanner_params.materialize_slots.size();
+        for (int j = 0; j < num_materialize_columns; j++) {
+            int index = _scanner_params.materialize_index_in_chunk[j];
             const Slice& field = fields[_scanner_params.materialize_slots[j]->id() - 1];
```
Comment (dirtysalt) [inline be/src/exec/vectorized/hdfs_scanner_text.cpp:169]: Can you be sure that index of `fields` is `materialize_slots[j]->id()-1`?

Comment (caneGuy) [reply be/src/exec/vectorized/hdfs_scanner_text.cpp:169]: @dirtysalt i think for text, if we do not store original column id int external table metadata. it is difficult to align.
result below is impala shows:
columns of test data is :
CountryID int , CountryName string, Capital string, Population string
```
create external table if not exists countries_list_1(Capital string, CountryID int ,  CountryName string, Population string) comment 'List of countries' row format delimited fields terminated by ',' stored as textfile location '/tmp/yuzhou';
```
```
[emr-header-1.cluster-259859:21000] default> select * from countries_list_1;
Query: select * from countries_list_1
Query submitted at: 2022-01-15 10:56:29 (Coordinator: http://emr-worker-3.cluster-259859:25000)
Query progress can be monitored at: http://emr-worker-3.cluster-259859:25000/query_plan?query_id=e5401f6bd049906f:8e80a59e00000000
+-----------+-----------+-------------+------------+
| capital   | countryid | countryname | population |
+-----------+-----------+-------------+------------+
| countryid | NULL      | capital     | population |
| 0         | NULL      | BeiJing     | 80         |
| 1         | NULL      | Wash        | 328        |
| 0         | NULL      | BeiJing     | 80         |
| 1         | NULL      | Wash        | 328        |
+-----------+-----------+-------------+------------+
WARNINGS: Error converting column: 1 to INT
Error parsing row: file: hdfs://emr-header-1.cluster-259859:9000/tmp/yuzhou/countries_1.csv, before offset: 75
Error converting column: 1 to INT
Error parsing row: file: hdfs://emr-header-1.cluster-259859:9000/tmp/yuzhou/countries_1.csv, before offset: 75
Error converting column: 1 to INT
Error parsing row: file: hdfs://emr-header-1.cluster-259859:9000/tmp/yuzhou/countries_1.csv, before offset: 75
Error converting column: 1 to INT
Error parsing row: file: hdfs://emr-header-1.cluster-259859:9000/tmp/yuzhou/countries.csv, before offset: 34
Error converting column: 1 to INT
Error parsing row: file: hdfs://emr-header-1.cluster-259859:9000/tmp/yuzhou/countries.csv, before offset: 34
```

Comment (dirtysalt) [reply be/src/exec/vectorized/hdfs_scanner_text.cpp:169]: Got it.

**Review by dirtysalt — DISMISSED** [review]

**Comment by caneGuy** [comment]
@imay could you help review this?thanks

File: be/src/exec/vectorized/hdfs_scanner_text.cpp [hunk 6]
```diff
@@ -197,4 +236,21 @@ Status HdfsTextScanner::parse_csv(int chunk_size, ChunkPtr* chunk) {
     return chunk->get()->num_rows() > 0 ? Status::OK() : Status::EndOfFile("");
 }
 
+Status HdfsTextScanner::_create_or_reinit_reader() {
+    const THdfsScanRange* scan_range = _scanner_params.scan_ranges[_current_range_index];
+    if (_current_range_index == 0) {
+        _reader = std::make_unique<HdfsScannerCSVReader>(_scanner_params.fs, _record_delimiter, _field_delimiter,
+                                                         scan_range->offset, scan_range->length, scan_range->file_length);
+    } else {
+        dynamic_cast<HdfsScannerCSVReader*>(_reader.get())->reset(scan_range->offset, scan_range->length);
```
Comment (dirtysalt) [inline be/src/exec/vectorized/hdfs_scanner_text.cpp:246]: Could you check `down_cast`?

**Comment by DorianZheng** [comment]
@caneGuy Please refine the commit message

**Comment by caneGuy** [comment]
> @caneGuy Please refine the commit message

done

**Review by dirtysalt — APPROVED** [review]

**Review by imay — APPROVED** [review]

**Comment by imay** [comment]
@caneGuy Please cherry-pick to branch 2.0 and branch 2.1

**Pull request merged**